### ISSUE TITLE
Companion-link service

### ISSUE THREAD

**Issue by hieusinh25** [postlund/pyatv#2515]
### What do you need help with?

- Can you show me what port companion client connect to companion link? Tks.

**Comment by postlund** [comment]
It's the port announced in the `_companion-link._tcp.local.` zeroconf service.

**Comment by hieusinh25** [comment]
oh. tks ! :3

**Comment by hieuttmatech** [comment]
can you write opack.py to swift code @postlund

**Comment by hieusinh25** [comment]
can you write opack.py to swift code

**Comment by postlund** [comment]
Unfortunately no, I don't have time for that (nor have I ever coded in swift). But the logic is pretty easy so should not be that hard. It's documented here as well:

https://pyatv.dev/documentation/protocols/#opack

**Comment by hieusinh25** [comment]
I have one question. In document of companion link service: https://pyatv.dev/documentation/protocols/#companion-link
. after 6 step of [Pairing](https://pyatv.dev/documentation/protocols/#pairing) then verification start ?. I mean when [Verification](https://pyatv.dev/documentation/protocols/#verification) start? . and after pairing success. in the next call pairing, command say: don't need to pairing. show just need verification in the next call

**Comment by postlund** [comment]
@hieusinh25 You need to pair once in order to exchange encryption related information and establish trust. Once pairing has finished, you save the keys you have obtained persistently. Whenever you want to connect, you perform verification as part of setting up a session (encryption) key for that connection. All data is encrypted with that key using Chacha20Poly1305 and the session key (as well as counters in each direction).

**Comment by hieusinh25** [comment]
so we only need pairing once time. and save the key. and for next connecting to Apple TV, don't need to pairing , just call verification ? 
I descript the step :
In first day:
- I pairing -> pairing done -> call verification. -> verification done -> call remote command -> close connection, turn off tv.
And in sencond day I wanna to connection tv again. So dont need pairing to because use saved key in first day. just need pairing before using remote command ? 
Please help me ? @postlund

**Comment by postlund** [comment]
Correct, but you will need to do verification "next day" when you set up a new connection to generate encryption keys for that session (based on that keys you created when pairing).

**Comment by hieusinh25** [comment]
tks @postlund . great helps for me!! 🥇

**Comment by hieusinh25** [comment]
def step2(self, atv_pub_key, atv_salt):

 the pramater put to func self._session.process(pk_str, salt). button seem atv_pub_key or pk_str dont use anymore. so the atv_pub_key isn't important. @postlund . can you explain me when use the atv_pub_key (return from self._atv_pub_key = pairing_data[TlvValue.PublicKey] when call PS_Start)?

**Comment by hieusinh25** [comment]
@postlund  I have one question please help me.

in connection() function of api.py . 
you have call some function below:

`        await self.system_info()
        await self._touch_start()
        await self._session_start()
        await self._text_input_start()

        await self.subscribe_event("_iMC")
        `
I have one question that these functions need to call ?
because after I pairing then verify succes  (get inkey, outkey to encrypt and decrypt using chacha20poly135)-> call turnoff or call some remote function -> nothing happened. 
in connection function dont call 

`        await self.system_info()
        await self._touch_start()
        await self._text_input_start()
        
        just call       `await self._session_start()` and this call success then I get _sid successful>
        
        Please help me. tks @postlund

**Comment by postlund** [comment]
You only need to setup a session and send system info, the remaining things are optional depending on what you want to be able to do.

**Comment by hieusinh25** [comment]
oh maybe I miss system info. because I only call setup a new session. and I got error:

tcp_input [C1:1] flags=[R] seq=3231763486, ack=0, win=0 state=CLOSE_WAIT rcv_nxt=3231763486, snd_una=3146035952
Connection failed: POSIXErrorCode(rawValue: 54): Connection reset by peer
tks you. So I will try to call system infor

**Comment by postlund** [comment]
I believe system info must be sent first of all, so yeah, that sounds reasonable.

**Comment by hieusinh25** [comment]
in   system_info(self) function have info . where I can get info variable. for the first time I connect to atv. how to get info to get   info.device_id, info.model, info = self.core.settings.info, info.name. or get it announced in the _companion-link._tcp.local. zeroconf service. ? @postlund . tks

**Comment by postlund** [comment]
I'm not really in a position to look that up right now, it's probably easiest if you just look at the code (you can also run `atvremote` with debug output and see what is sent). Otherwise you'll have to wait until I have some time.

**Comment by hieusinh25** [comment]
tks you @postlund

**Comment by hieusinh25** [comment]
I have call two function 
   await self.system_info()
    await self._session_start()
    
    But only self.system_info() return response. the self._session_start() don't return response and then I call remote command -> nothing happen. can you suggest me something ?

**Comment by hieusinh25** [comment]
tks @postlund . I find the bugs that I dont increate counter when descrypt . tks you a lot @postlund .
